### ISSUE TITLE
Display profile not selectable from system tray menu

### ISSUE THREAD

**Issue by erbkaiser** [terrymacdonald/DisplayMagician#49]
**Describe the bug**
My 'three screens' display profile is not possible to select from the DM system tray menu option, it is always grayed out. The profile does work if selected from the Display Profiles page in the program or from a shortcut.

**To Reproduce**
Steps to reproduce the behaviour:
1. Have three screens on an nVidia graphics cards that supports Surround
2. Create a 'Surround' profile with all three screens treated as one monitor: 'Span displays with Surround' inside the nVidia Control Panel.
3. Create a 'Three screens' profile with all three screens treated as individual monitors.
4. Change to the Surround profile, then fruitlessly attempt to swap back to Three Screens from the systray

**Expected behaviour**
Since swapping FROM Three screens to surround works, I expect the other way to also work.

**Screenshots**
Display profiles:
![Three screens](https://user-images.githubusercontent.com/6014923/138554254-0fbecbc2-8532-4aca-961a-5f6c79573188.png)
![Surround](https://user-images.githubusercontent.com/6014923/138554257-25b599f8-ea3a-42d6-b340-735bbc93a5fb.png)

System tray behaviour:
![Systray](https://user-images.githubusercontent.com/6014923/138554255-af17347c-2f31-4a15-9e3b-776cd4e237cc.png)

**Reporting Logs**
[DisplayMagician.log](https://github.com/terrymacdonald/DisplayMagician/files/7402813/DisplayMagician.log)

**Environment (please complete the following information):**
 - Windows Version: Win11
 - DisplayMagician Version: 2.0.1.59
 - Video Card: Gigabyte nVidia GeForce RTX 2060 Super
 - Number and make/model of monitors: 3x Samsung SyncMaster P2250
 - Video Card Driver Version: 496.13
 - Date of last video driver update: 2021-10-12
 - Date of last windows update: 2021-10-23

**Additional context**

**Comment by erbkaiser** [comment]
Thanks for your reply.

To clarify, I can always switch to the 'Three Screens' profile by using the Display Profiles from the DM program, or by using a saved desktop shortcut for the profile. But the system tray option never works, no matter which of my three profiles I have active.

To perform the test, so I deleted the Three Screens profile and saved it again as aaa Three.
The systray menu still showed 'Three Screens' so I closed DisplayMagician and restarted it.

Now, I am seeing 'aaa Three' as the first item in the list, but it is again grayed out. So it would appear to be something unique to this display profile itself.

![aaa three](https://user-images.githubusercontent.com/6014923/138574771-9cdb431e-498f-418b-b56b-6e89127ed862.png)

**Comment by terrymacdonald** [comment]
Woah, this is much stranger again! Ok so the potentially broken games are
all Steam games? So you have any Epic, Origin or GOG games to test as well?

I'll check the desktop launching late today just to see if it's a bug I can
replicate too.

What version of Steam do you have? If it's an error with how DM works with
Steam then we're potentially narrowing it down.

We know it's not a missing all, and it's not a missing .net framework. We
know it might be related to Steam games in some way. Or games with long
names....

Ok, did you say you're using windows 11? Cause I wonder if there is an
issue with how the notification window works in windows 11? I wonder if
it's a bug in the windows 11 side? Or if it works differently in some way...

Can you attach a DisplayMagician.log file from one of the failed Steam
games launches with the long names to the GitHub issue please? It will give
me more information as to where and error might lie. I log any problems
with error messages and I think I instrumented that part of the codebase
write well from memory.

I'm still on windows 10, and I might need to upgrade to windows 11 in order
to test how it works on there. That will take a few days for me to get time
to do.

Thanks
Terry

**Comment by erbkaiser** [comment]
My current Steam version is Oct 13 2021's, the current one as of this time.

One possible cause for the initially broken shortcuts that comes to mind is that I have a large library of 6400+ games. Looking at the log, DM seems to try to load every game I possibly own. After launch the log file will grow for about 3 minutes to around 32000 lines. 31696 in the one provided.

It will stop growing there unless I open the Game Shortcuts window. Every click in that window causes the file to grow again. I did not test to see how large it could get.

Included is the (zipped) log file from launching 'Deathtrap Dungeon: The Interactive Video Adventure' - the game with the long name that causes a crash I don't see for other games. Note that I had to manually close the DisplayMagic window as it doesn't recognise the game has stopped playing, following the crash.
The game itself works fine by the way.
[DisplayMagician.zip](https://github.com/terrymacdonald/DisplayMagician/files/7444729/DisplayMagician.zip)

DisplayMagic stops responding to the system tray menu following the crash, so for the rest of the tests I quit it with task manager and restarted it.

x-x

Children of Morta is a game I own on GOG. The desktop shortcut and systray launch work fine for it.

Hades I own through Epic, it also works fine with shortcut and systray.

Note that I didn't let DM swap display profiles for these, as that is a function that always works anyway.

Log file included for all four launches. 
[Morta and Hades.zip](https://github.com/terrymacdonald/DisplayMagician/files/7444766/Morta.and.Hades.zip)

Conclusion: the problem with game shortcuts not working from systray or desktop seems limited to Steam on my machine. And I know I am an outlier in the size of my library.

I think the long game name crash is an actual bug since it references a fixed size, and it does not happen for shorter games but does happen for Deathtrap Dungeon as well as another game (demo),  "Dr. Langeskov, The Tiger, and The Terribly Cursed Emerald: A Whirlwind Heist" -- specifically chosen since it's the longest name I know of.

**Comment by terrymacdonald** [comment]
Wow. That's the largest game library I've ever heard of! I have about 30
games by comparison 😁

The crash for the deathtrap dungeon and Dr. Langeskov I think are related
to the special characters. I'm not sure where exactly, but that's my gut
feel. It could be that the special characters are causing issues when used
in the . I'll try finding a game with a colon in its name and I'll try to
add it to my Steam games and see if I can make it crash.

I've also just upgraded my development machine to windows 11 so I can test
there and see if that introduces any issues.

Thanks for all the troubleshooting, it's helping a lot.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@erbkaiser So I've found and fixed the long character issue. You were indeed correct, as there was a logic bug which meant the notify icon took more than the maximum allowable 63 characters. I've fixed that in the codebase, and I'll shortly look at how I can update the shortcuts/profiles in the system tray as soon as they are made.

Then after that I'll look at the rest of the many bugs you're finding. This is all excellent news, as I'd rather have them found and fixed before I do a wide release to lots of people. It's much better to have a select group of testers using early release builds and then ironing out any issues before they hit mainstream. SO THANK YOU! 👍 

Once I've fixed the next round, I'll make a new test release up for you to install, and then you can check to see what's fixed, and what I need to repair. 

Oh - I was trying to figure out how to speed up the loading with so many games are involved, but I have to parse through them all on start up in order to figure out which ones are installed. I'm already doing that in a thread per game, so I was wondering if it was just worth providing a pop up window mentioning something like 'DisplayMagician is loading your games, and if you have lots this may take a while' or something like that? What do you think?

Thanks
Terry

**Comment by terrymacdonald** [comment]
@erbkaiser So here is a new build that will hopefully fix some of the issues you've got. Please install using the included setup exe. 

For the steam related games not showing up, I wonder if you've got so many games in the steam library database that we parse, that the thread it's running in is timing out? I'll have to delve into your logs that you provided. At this stage I've been fixing the issues that you've described as I've found them when running the same tests myself.

I've fixed the following:
- Shortcuts/profiles now update as soon as they are saved
- Long named games now don't crash DisplayMagician :)
- There is now a splashscreen on startup to tell users that DisplayMagician has started and is loading games, and that it may take ages if you have lots of games (user can turn off in the settings form).

Here is a new build with those fixes in it. Can you please the issues you've found in the past, and list the bits that are still broken? I'll then spend some time tomorrow fixing those as well!

[DisplayMagicianSetup-v2.1.0.57-test.zip](https://github.com/terrymacdonald/DisplayMagician/files/7445952/DisplayMagicianSetup-v2.1.0.57-test.zip)

Thanks
Terry

**Comment by erbkaiser** [comment]
Thank you for putting so much time in trying to fix all the weird issues I'm finding!

Uninstalled DisplayMagician, deleted the folder in program files, then installed the new build.

A lot of improvements :)
The long game name crash is fixed, as is the systray behaviour (mostly, see issue 2)

x -x 

Remaining issues I've found:

1) The splash screen disappears quickly, but Steam game profiles are still greyed out in the systray. If I open the DisplayMagician Game Shortcut Library immediately after, sometimes the Steam games still show a red error and this fixes itself the moment I close that window. Other times it's fixed by the time I open the window. Once it does, both library and systray are working for Steam games.
I tried waiting it out but this does not seem to allow the systray to update -- but swapping display profiles or launching a non-Steam game shortcut do.

2) Display Profile switching from the systray is still unreliable. One option is greyed out, usually the active profile but if I switch profiles using a game shortcut the one the game shortcut switched TO is greyed out, even after returning.
Example: I start in 'Three Screens' and launch a game with 'Surround', then after closing the game 'Surround' is greyed out.
If I manually switch from TS to Surround, before the switch TS is greyed out and after, Surround is (functional, is this expected behaviour?)

3) Steam game DisplayMagician desktop shortcuts still don't work. There is also a usability quirk with the error and the splash, it appears overlapped at first 
![image](https://user-images.githubusercontent.com/6014923/139526548-ac2e6dd5-f3c3-49f6-92e9-61790eb48077.png)
This is what it reads:
![image](https://user-images.githubusercontent.com/6014923/139526464-d6609aaf-fec2-43b6-b91d-0d7bcbca2766.png)
After dismissing the "can't be launched" error, the "DisplayMagician is loading..." splash screen remains active. I have to close it with task manager.
Disabling the splash in settings does not allow the games to launch.

4) The splash screen appears if I switch display profiles from a desktop shortcut and doesn't go away until I close it with task manager. Even if DM is running already.

5) The game profile for American Truck Simulator (Steam) should (still) launch two additional programs, it doesn't. Additionally, DisplayMagician still crashes after closing it if these programs were not started manually before I launched the DM shortcut for the game. 
[ats after closing.txt](https://github.com/terrymacdonald/DisplayMagician/files/7446139/ats.after.closing.txt)

This crash also causes DM to not return to the previous display profile, exhibiting issue number 2 here.
Log for this session included. I start in Three Screens and the game goes to Surround, and does not switch back: 
[DisplayMagician.zip](https://github.com/terrymacdonald/DisplayMagician/files/7446141/DisplayMagician.zip)

It's not all Steam games with extras as I've set 112 Operator to open Notepad++ before launching, and that works fine. Notepad++ launches as the game does and is properly closed after. Log for this session included. I start in Surround, the game goes to Three Screens, and I return to Surround. [112 operator.zip](https://github.com/terrymacdonald/DisplayMagician/files/7446152/112.operator.zip)

The same ATS game with two different programs on prelaunch (Notepad++ and cmd.exe) also works fine.

After some more testing I found it has to be these specific extra programs I'm launching. if I apply either Local Radio Server or TrackIR v5 to any other game I get the same errors there, where the apps won't launch. I can only guess to the cause -- it's not that Local Radio Server requires elevation as any other program that needs this will work.

**Comment by terrymacdonald** [comment]
Woah, that's some good testing @erbkaiser! Very detailed feedback. I'll
with through some of those today.

My gut feel is that the splash screen issues should be easy enough to fix,
but the Steam one is odd. It's almost like it's giving up before it's
finished. I'll have to take a deeper look into the Steam part of the logs
to see if there is any sort of indication what's going on.

I think we'll eventually sort all of the issues so it will just work for
you, it just might take a few more sessions of building and testing until
we done. Hope that's still going up be ok?

Thanks
Terry

**Comment by erbkaiser** [comment]
I'm happy to keep testing for as long as you want to :)

**Comment by terrymacdonald** [comment]
OK @erbkaiser,

I've made a lot of changes to the code to try and remove most of the issues you've highlighted. I'll go through them one by one.

1. I think the greyed out steam icons might actually be due to the fact I made the game loading parallelised. My focus was on trying to get to the main screen as quickly as possible, but that in turn meant that if there were a lot of games, then they may not be loaded in time for the MainForm to open. The MainForm also didn't have any logic to actually get notified of the games finishing getting loaded, so it wouldn't update itself to get the full list. The simplest fix (and the one I've implemented) is to just wait until all the games are loaded. Now that we have a splash screen that becomes an option. 

I'm still not 100% confident that this is the final fix, but hopefully it allows enough time for parsing your massive list of games :). I'd really like to know how long it takes now for DisplayMagician to load i.e. how long does the splash screen show on startup?

2. To make things clearer, I've now changed the profiles part of the notification right-click menu to show the currently active profile in bold. This should make it a lot clearer what is happening. 

Additionally, I've moved where the right-click update logic is, so it now sits in each of the buttons on the DisplayProfile page and the Shortcut library page. This means that things are only updated when you click on the buttons that result in them being updated. This again should make the whole update process more reliable. 

3. The splash screen sets itself as topmost now when it starts, but then straight away turns that off again. This should have the effect of putting the window in the front, but then allowing other windows (like error messages to go over the top of the splash screen.

The second part of that error hasn't been fixed yet, as I wanted to clear up the rest of the errors first, and then see what is left. There is a change that the Steam game loading issues will have an impact on this error and may fix it... that's what I'd like you to test next :).

4. Should be fixed now. I think it was related to the error messages stopping execution, which didn't allow the splash screen removal logic to work. I reversed the order of the logic so that the splash screen should hide before the error messages show now.

5. There was some new process monitoring logic introduced in v2.0.1 which tried to be much smarter in how DisplayMagician handles monitoring programs. Some applications run from a single exe. Others have a 'loader' that actually loads something else then exits. DisplayMagician now tries to figure out what is happening, and tries to close things the right way when a game exits. But I expect there to be bugs in this process as it is just so darn complicated to do!

One thing I did notice is that the start program closing logic didn't have any 'safety nets' around it, so when an exception occurred then the rest of the profile reversing wouldn't take place. I've now added those safety nets to that bit of code so hopefully even if a program can't be stopped then DisplayMagician will still change back to the original display profile.

---

So, as you can see, lots of changes, and hopefully it clears away most of them so I can fix the last few remaining ones!

Please install this new build, and do the same awesome testing you've been doing! It's REALLY helping me out, so thank you.

[DisplayMagicianSetup-v2.1.0.71.zip](https://github.com/terrymacdonald/DisplayMagician/files/7447645/DisplayMagicianSetup-v2.1.0.71.zip)


Thanks
Terry

**Comment by erbkaiser** [comment]
1) I have DisplayMagician set to launch with Windows and it now takes a while before the tray icon appears. Once it does, everything seems to be there including Steam games. I can see it's running in task manager but there is no splash screen during initial load after logging into Windows -- I don't mind this.

2) The bold for active profile is a nice touch, much clearer than greying it out. Adding a new display profile or game shortcut is now immediately reflected in the systray menu as well.

3) **Steam desktop shortcuts are now working!** 

4 and 5 I could not test as the errors are fixed and simply don't re-appear.

I did find one new issue/reversion: DisplayMagic no longer pulls the game icons for games from launchers (Steam, Epic, Uplay, GoG -- I don't have EA installed). Instead, it shows the launcher icon. 
![image](https://user-images.githubusercontent.com/6014923/139595846-6e7b472d-66c7-43ef-952e-e457b0160e3a.png)
Existing game shortcuts kept their icons until I updated them.

It does show icons for some games in the window where I add a new shortcut, but even for those where it does, it won't appear on the final profile.
![image](https://user-images.githubusercontent.com/6014923/139595839-fcc9fa0f-50f1-4b6a-b612-67197accc1f9.png)
That it lists some games double or triple is something that I think is caused by GoG Galaxy mirroring my other libraries, as I am pretty sure I didn't see it when I had uninstalled Galaxy a while back. I didn't think to mention this before.

Desktop game icons never really worked reliable, in DisplayMagician 1.0 it showed the screen icon for me, with 2.0 builds up until this one it sometimes showed a greyscale icon, and now it shows the launcher, so I consider this a very minor issue and am just reporting it since it is a change.

All in all a major step forwards. At this stage the only remaining issue I can find are the problem with the two programs that will not launch with the game ATS -- easily worked around.

**Comment by terrymacdonald** [comment]
Great news!

The double up on icons was a mistake on my part with the version I
released. I realised I had a separate thread that would then run the game
parsing in their own threads individually. I changed this and accidentally
added the game loading twice. It's already fixed in the codebase.

The icons thing is concerning. I'll have to attend a bit of time checking
each library to make sure it's pulling the icons from the correct places.
Icon handling is so difficult as there is no standard places for the icons
and no standard format for them. DM used about 4 different methods for
trying to extract icons, so it's annoying to me that is not working as well
as it should. I'll drop some research and see if it can fix that.

I might add the ability to set a manual icon override, as that would at
least people manually pick the icon they want if DisplayMagician gets it
wrong

Again, thanks for all you help testing. We've found and fixed many bugs and
it's really been amazing. This will make DisplayMagician so much better!
You are a rock star.

Thanks
Terry

**Comment by terrymacdonald** [comment]
ok @erbkaiser, So I basically rewrote the bitmap and icon handling routines tonight to smooth them out and hopefully make them more robust. I think they're faster too actually.... but I also added manual icon overriding! What this means is that you can choose an icon file or an exe that has better icons in it, and that's the icon DisplayMagician will use for the shortcut library and the desktop shortcut. Works quite well IMHO. But I'd love you to test.

[DisplayMagicianSetup-v2.1.0.91.zip](https://github.com/terrymacdonald/DisplayMagician/files/7452037/DisplayMagicianSetup-v2.1.0.91.zip)

There should be a splash screen now on startup. It is possible that the splash screen is behind another window on start up and so it won't be shown. It does have a splashscreen on my machine, so if it doesn't on yours then we'll need to do a bit more testing.

The two programs that won't launch as start programs is really puzzling me! Can you please create a screen shot showing the start programs screen in DisplayMagician for ATS? I might be able to see something that will give me an indication of what is wrong.

In any case, please test DisplayMagician 2.1.0.91, and then we'll see what bugs are left after this! 

Thanks
Terry

**Comment by erbkaiser** [comment]
Thanks for the new build.

The splash screen now works as expected, when it disappears DisplayMagician is ready to load.

Automatic icon selection for games is also fixed -- I had to edit existing profiles.

I have a request for the icon override - can you make it possible to select a specific icon inside a DLL or exe? With the truck games, the exe has 25 icons in it, of four different sizes. Three of these sets are identical and in colour but the first set is in greyscale and that is what DM ends up using.

In trying to extract one of these icon files myself I managed to catastrophically break DisplayMagician:
![image](https://user-images.githubusercontent.com/6014923/139651087-5cce8933-d6a1-49d1-b0ba-c08ec294ca3d.png)
Opening the game shortcuts window causes the error to appear. This is unrecoverable within the program itself.
What I did was extract the raw image data for one of the colour icons from the exe, save it as JPG, and rename that to ICO. DM would not let me select it by default so I tried to force it by changing the selector to all file types. Bad idea :D 

Here's the crash log 
[incorrect icon crash.txt](https://github.com/terrymacdonald/DisplayMagician/files/7452152/incorrect.icon.crash.txt)
 And the shortcuts file from Local Appdata 
[Shortcuts_2.0.zip](https://github.com/terrymacdonald/DisplayMagician/files/7452154/Shortcuts_2.0.zip)
I fixed it by removing the ATS shortcut section from that file. Just editing the UserChoseOwnIcon/UserIconPath did not work.
Here's the broken icon file I tried to make [ats.zip](https://github.com/terrymacdonald/DisplayMagician/files/7452199/ats.zip) - setting it as the icon for any game will cause the window to break catastrophically.

Re-created the profile in a normal state and here it is with both programs listed: 
![image](https://user-images.githubusercontent.com/6014923/139653080-caaab6a0-76c8-4f3f-926e-dacb74cf8061.png)
It will launch the main amtrucks.exe fine through Steam but refused to launch these two addons. So far, every other program I tested launches fine. The error is not related to the game as any other game, including a non DRM one, will also never launch either of these two.

**Comment by terrymacdonald** [comment]
Yeah, icon files are their own particular type of file format. Renaming a
JPG definitely won't work 😁. It will break things for sure. Interesting
though that it results in a crash to desktop. I'll have to see if I can
stop that happening somehow.

Good idea about an icon picker, though that's actually quite a bit of work.
I'll have a look tonight, but I might need to leave that for 2.1.1 in order
to get this version out the door.

Terry

**Comment by terrymacdonald** [comment]
@erbkaiser So the icon picker idea piqued my interest, so I started implementing it tonight. It's coming along really well, but as mentioned earlier it's a bit of work. I reckon I'll get it done by the end of the week, and I think it's a much better solution that I had before. It's WAAAAAY easier to get the images you want to use as the game bitmap/icons. 

Hopefully this all works out, but it's looking good at the moment.

Terry

**Comment by erbkaiser** [comment]
Excellent, looking forward to it!

**Comment by terrymacdonald** [comment]
@erbkaiser Great news. I've finished the shortcut image selector updates. Can you please take a look and make sure the image selection stuff is working for you. Please note I split the graphics loading bits into two parts in an effort to make the loading process faster for when running games, and traded that off with having the editing/creating shortcut window loading slightly slower. I'm just about to add another 'please wait' message to the shortcutForm to warn people such as yourselves with massive game libraries that it will take a little while to load the shortcut editing form.

Anyway - can you please test, and note any issues you find (as you have been doing).  You should be able to install this test version 2.1.0.160 over the top of your last install. 

[DisplayMagicianSetup-v2.1.0.160.zip](https://github.com/terrymacdonald/DisplayMagician/files/7480411/DisplayMagicianSetup-v2.1.0.160.zip)

Thanks!
Terry

**Comment by erbkaiser** [comment]
I did some testing: this version is great.

Shortcut creation/editing is a bit slower, it takes me about 4 seconds before the Configure Game Shortcut window pops up after clicking edit.

The icon selection works: creating a new shortcut seems to default to a better image now in some cases and for the rest I can freely select from any images included.
Selecting an image file or a different exe to add images to the list also works. I tried with exe, JPG, and PNG - great to being able to set correct images for those few games where the default icon selection still fails.

I encountered no errors in launching game shortcuts, swapping profiles, creating new profiles and shortcuts.

The only remaining issue is the weird one with the two apps not launching which is a minor one for me.

**Comment by terrymacdonald** [comment]
Hi,

Can you please send me some links to where I can download those two bits of
software please? I have a suspicion that my process launching logic doesn't
work for whatever mechanism those two programs use, so I'd at least like to
install them to test how they work.

Thanks
Terry

**Comment by erbkaiser** [comment]
ETS Local Radio can be downloaded here https://koenvh.nl/projects/ets2-local-radio

TrackIR v5 on https://www.naturalpoint.com/trackir/downloads/ -- it won't do much without a head tracker, but it should launch.

**Comment by terrymacdonald** [comment]
@erbkaiser so after checking in detail how the start programs are being started, I'm actually surprised that it's working for any start programs. It's not very robust code, so I'm going to rewrite it and improve it greatly. There are a few robustness things I want to do to make sure it can cope with the many different ways that programs can be started, so it is likely to take quite a few days before I have it the way that I want it.

But, this should be the last few changes I need to make before it's ready to ship!

Terry

**Comment by terrymacdonald** [comment]
@erbkaiser As seems to commonly be the case, I'm a faster programmer than I thought I was! ;D

I've now got a version of DisplayMagician that I think is pretty robust. I've been able to find some amazing example code that I've tweaked, modified and extended to completely rebuild the process starting library. The new process library does a few cool things:

- It tries to launch programs with the correct process priority, meaning that you can select a higher priority than default if you want just a few more frames per second
- It tries to catch launcher processes, and then automatically finds their child processes and monitors them instead.
- It handles multiple child processes if needed
- If it can't use some of the lower-level win32 library calls, it falls back to the standard C# process library calls, so the programs at least get started (even though they may not have a custom process priority)

I'm pretty darn happy with the code that came out of the rewrite process now, and I feel like your help pointing out all the broken bits really helped pull DisplayMagician forwards. It feels like it's **nearly** ready to ship (though I might need some more tweaks to support NVIDIA Display Config extraction...that's another story).

Anyway, can you please test this latest version, and see if there is anything that doesn't work. I'd love to get all the issues fixed if I can :D.

[DisplayMagicianSetup-v2.1.0.197.zip](https://github.com/terrymacdonald/DisplayMagician/files/7491913/DisplayMagicianSetup-v2.1.0.197.zip)

Thanks
Terry

**Comment by erbkaiser** [comment]
The new splash screen before editing game profiles is a good addition.

I haven't found any issues at all. Even TrackIR5 and Local Radio now launch before the game does. I've thrown a few programs at it including Modern UI apps, and everything just works.

Excellent work, thank you so much!

**Comment by terrymacdonald** [comment]
Hi,

I've had another person say that they have two games listed in the list of
games within the shortcut editing window. Can you please see if you have
that?

Also, they say that they can't run a game shortcut with a surround display
profile. It swaps the displays before it launches, but then exits.

Can you please check that for me too?

Thanks
Terry

**Comment by erbkaiser** [comment]
I indeed still have many Steam games listed twice in the game shortcut window. It doesn't duplicate all of them but maybe more than half.

Launching an nVidia Surround profile works for me, I did recreate display profiles and set the game shortcuts to the new profiles a few builds back.
I get the splash screen that DM is loading (even if it already is), displays are swapped, the game (and associated programs) launch. And when I close the game the original display profile (three separate screens in my case) is restored.

**Comment by terrymacdonald** [comment]
Perfect. Thank you for testing that, it clears things up for me and gives
me an idea of where the error could be happening!

👍

Terry

**Comment by terrymacdonald** [comment]
Hi @erbkaiser,

We've recently found and fixed an error with the Windows Library and NVIDIA Library that I made. It would come up with an error when it was validating a config, and that would stop the windows portion of the display config being applied. This is now fixed.

Additionally, we had an issue with the NVIDIA Library not correctly applying colour information and failing if a screen with colour information was turned off in a profile. This too has been fixed.

And lastly, the doubled Steam information is fixed. I was accidentally adding the base steam library to the config twice, so all games in the base game library were being scanned twice (doh!). This has been fixed too.

Could you please update your DisplayMagician and test this version? [DisplayMagicianSetup-v2.1.0.215.zip](https://github.com/terrymacdonald/DisplayMagician/files/7531681/DisplayMagicianSetup-v2.1.0.215.zip). I think most of the bugs are now squashed, so I'm sending this copy to my faithful testers to doublecheck the reliability on their own computers.

IMPORTANT: I strongly recommend deleting all your existing Display Profiles and recreating them again. This will ensure that your Display Profiles have all the latest fixes applied to them. Once you recreate the Display Profiles you can then go through your shortcuts and update them to use the new profiles.

Thanks
Terry

**Comment by erbkaiser** [comment]
After installing the new build and starting it, DisplayMagician would crash on load with a 'Fatal Error. Value cannot be null.'
I had to rename (or remove) the DM folder in %LocalAppData% to get it to work. I saved the folder so if the log file can help, here it is:
[DisplayMagician.log](https://github.com/terrymacdonald/DisplayMagician/files/7533594/DisplayMagician.log)

-x-x-

In a clean profile, creating display profiles works fine but there is an issue with shortcuts.  I can create one game shortcut fine, but trying to create a second seems to open the properties of the existing one again. Editing this to a new game corrupts the existing shortcut and makes it a clone, with a broken icon.
When I tried to delete one, DisplayMagician crashed with 'Exception of type 'DisplayMagician.ShortcutRepositoryException' was thrown.'

This does not APPEAR to happen if I exit the Game Shortcut Library and re-open it before creating additional shortcuts, but even then shortcuts will be broken.
It is reproducible by these steps:
-Open the Game Shortcut Library and add a new shortcut, or edit an existing one.
-Save the shortcut
-Add a second shortcut, or edit a second existing one
-Notice that the properties of the first opened shortcut are shown.
-If the second shortcut is saved, both shortcuts appear corrupted.

If the GSL is closed after creating the first shortcut and before creating the second one, the bug re-appears the first time I try to edit either of the two existing shortcuts.

There is only one shortcut icon file  saved in the %LocalAppdata%\DisplayMagician\Shortcuts folder.

Game shortcuts saved to desktop will all take the icon of the first profile in the list, but will take the name of the game they were created for. They will also launch the first game.

x-x

When I replaced the new Shortcuts folder contents with the old one, existing shortcuts work fine including desktop shortcuts, but any NEW shortcuts exhibit the same broken behaviour.

In this case, only one new (broken) shortcut icon file would be saved to the Shortcuts folder with the old ones working still.

**Comment by terrymacdonald** [comment]
Darn it. Not sure how I missed that one in my testing either. Sorry. I'll
go fix and give you any updated one in a day or so.

Thanks
Terry

**Comment by terrymacdonald** [comment]
OK - Fixed! Had to change way more than I thought I would, but it the UI *should* work now. Can you please test DisplayMagician v2.1.0.239 and see what happens. 

[DisplayMagicianSetup-v2.1.0.239.zip](https://github.com/terrymacdonald/DisplayMagician/files/7537310/DisplayMagicianSetup-v2.1.0.239.zip)

I am very interested in whether you can successfully swap to a triple screen profile after swapping to a surround screen profile. Some other testers have a problem with that, but I don't on my test system. I'd like to know if you do.

Thanks for all your help @erbkaiser!

**Comment by erbkaiser** [comment]
Hi Terry,

thanks for the new build. This fixes the game shortcuts issue.

Switching from triple screen to surround and back works, I can switch between all display profiles without problems. 

I didn't notice any issues in my tests of the program.

**Comment by terrymacdonald** [comment]
Perfect! That's what I was hoping for 😁. Thank you so much for the tests.
It's so valuable to me.

Thanks
Terry

**Comment by terrymacdonald** [comment]
Hi @erbkaiser ,

So domenic and I found some errors in the Windows CCD WinLibrary which would affect you as an NVIDIA user (see #41). We worked over the last week and a half to get the bugs squashed, and I've released them in a new test build of DisplayMagician here: [DisplayMagicianSetup-v2.1.0.254.zip](https://github.com/terrymacdonald/DisplayMagician/files/7569245/DisplayMagicianSetup-v2.1.0.254.zip)

This is very nearly production ready as I've run out of any more bugs to squash. So can you please test this pre-release version to make sure it's fit for release? 

**IMPORTANT: If you have moved to WIndows 11, please delete all your DisplayProfiles as you need to recreate them if you moved to Windows 11 (each DisplayProfile is specific to the operating system it was made on). Then change all your game shortcuts to use the new DisplayProfiles.**

I'm hoping that once you've recreated your DisplayProfiles and updated your Game Shortcuts then everything will work as it should 🤞 

Good luck!

Thanks
Terry

**Comment by erbkaiser** [comment]
Hi Terry,

No bugs found! I've been trying to break DisplayMagician today but it withstood all tests.

I was already on Windows 11 since I first contacted you, but either way I recreated my display profiles and game shortcuts from scratch and have been trying everything I normally do, as well as some things I have never tested before like launching a program afterwards. DisplayMagician works perfectly. No crashes, no strange things. Steam, Gog Galaxy, Epic, Ubisoft Connect shortcuts are all detected by the program, no doubling, and launch as expected.

Thanks for all the work on this program!
